Lib/test/.ruff.toml

```
@@ -0,0 +1,38 @@
+fix = true
+select = [
+    "F811",  # Redefinition of unused variable (useful for finding test methods with the same name)
```

> > LGTM, thanks! I guess we just have to wait for #109139 to be merged first?
> 
> Yes, or exclude `test_monitoring` here and re-include it #109139.

This got me thinking, "Hmm... or we could just `# noqa` the problematic tests for now", which then got me thinking, "Hmm... if there's a possibility that we might have `# noqa` comments in `Lib/test/`, maybe we should add some validations for those in CI?"

```suggestion
    "F811",  # Redefinition of unused variable (useful for finding test methods with the same name)
    "RUF100",  # Unused noqa comment
    "PGH004",  # don't use blanket noqa comments; only ignore specific rules
```

---

PGH004 I'd be less sure about, as it ties NoQA comments to Ruff's format. For example I believe support is being introduced for the more readable pylint-like style codes, but these might then sneak in over the "standard" pycodestyle codes.

---

I've no strong opinion, but I don't think there's really a defined standard for `noqa` comments. It used to be "whatever pycodestyle/flake8 support", since it _used_ to be that these comments were specific to those tools. Now it's more like "whatever pycodestyle/flake8 support if you're using those tools, OR whatever pylint supports if you're using that tool, OR whatever ruff supports if you're using _that_ tool". We're using ruff in CI, not any of those other tools, and it makes sense to me to tie the noqa comments we'd be using to the tool we're using them for?

---

Fair enough, I suppose the concern was potential churn/lock-in when the next great tool comes around, but not a blocker in any real sense.

---

I generally think these are a good idea, we have the Flake8 equivalents for Pillow.

But let's leave it out of this PR because there's quite a few warnings to fix:


<details>
<summary>PGH004 and RUF100 warnings</summary>

```
Lib/test/test_asyncio/test_events.py:2885:22: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_asyncio/test_events.py:2885:22: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_asyncio/test_runners.py:182:60: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_asyncio/test_runners.py:182:60: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_binascii.py:279:9: F811 Redefinition of unused `test_hex_roundtrip` from line 236
Lib/test/test_context.py:454:19: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_context.py:454:19: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_inspect.py:2798:37: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_inspect.py:2798:37: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_inspect.py:2801:54: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_inspect.py:2801:54: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_inspect.py:2806:43: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_inspect.py:2806:43: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_inspect.py:2810:44: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_inspect.py:2810:44: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_monitoring.py:1211:9: F811 Redefinition of unused `test_line_then_instruction` from line 1188
Lib/test/test_monitoring.py:1216:9: F811 Redefinition of unused `test_instruction_then_line` from line 1193
Lib/test/test_tomllib/burntsushi.py:11:116: RUF100 [*] Unused `noqa` directive (non-enabled: `E501`)
Lib/test/test_tomllib/burntsushi.py:12:104: RUF100 [*] Unused `noqa` directive (non-enabled: `E501`)
Lib/test/test_tomllib/burntsushi.py:22:20: RUF100 [*] Unused `noqa` directive (non-enabled: `C901`)
Lib/test/test_zipfile/__main__.py:3:27: RUF100 [*] Unused `noqa` directive (non-enabled: `F401`)
Found 21 errors.
[*] 11 potentially fixable with the --fix option.
```

</details>

And we'd have to decide _how_ to fix these, because we can see, for example, Ruff wants to strip this F401 "unused import" warning:

```diff
-from . import load_tests  # noqa: F401
+from . import load_tests
```

But we do "need" it, because it would normally be warned, even though Ruff has been configured to ignore F401 so thinks it's unused.

---

Good points. And various core devs often run pyflakes manually on the `Lib/` dir on a periodic basis in order to find unused imports. So these shouldn't be added for now, agreed!